### PR TITLE
Parse versions

### PR DESCRIPTION
**Pull request by EPashkin**
WIP

### CODE REVIEW

**Comment by EPashkin** [comment]
Stuck on:
1. Don't found way to early returns from .map() so `from_str` has manual iteration
2. Maybe same in `read_function`, don't have time to fully test.
3. Doc tests not compiling, not found why.

**Comment by EPashkin** [comment]
4. Found error in gir: `Version: 3.0.`. May be it same error for you add `.filter(!s.is_empty())` in gtk

**Comment by EPashkin** [comment]
Ignore fixing part.
But i still not fully understand for what you do it.

**Comment by gkoz** [comment]
I used `take_while` to cut of the iterator on a parsing error.
`filter(!is_empty())` in main crates' build scripts is needed to avoid returning an empty `cfg` that causes an ICE in rustc.

Regarding the version numbers complexity. Do you think there's any harm in using a tuple and defaulting failed elements to 0?

Then there're some strategies of acting on the versions to choose from: 
 * we may get away with ignoring the "patch" version (third element) and just leaving the door open to adding it if there's demand;
 * collect the list of x.y.z versions that are present in such attributes and put the list into the build scripts, so e.g. the build script would enumerate `pango_1_30`, `pango_1_31_0`, `pango_1_32`, `pango_1_32_4`, `pango_1_34`, `pango_1_36`, `pango_1_36_7`.

File: src/parser.rs [hunk 1]
```diff
@@ -593,6 +593,13 @@ impl Library {
         let name = try!(attrs.get("name").ok_or_else(|| mk_error!("Missing function name", parser)));
         let c_identifier = attrs.get("identifier").or_else(|| attrs.get("type"));
         let kind = try!(FunctionKind::from_str(kind_str).map_err(|why| mk_error!(why, parser)));
+        let version = match attrs.get("version") {
+            Some(v) => {
+                println!("Version: {}", v);
+                Some(try!(v.parse().map_err(|why| mk_error!(why, parser))))
+                },
```
Comment (gkoz) [inline src/parser.rs]: The Rust style asks us to align the closing brace one tab to the left (probably a typo?) and, surprisingly, not add the trailing comma.

Comment (EPashkin) [inline src/parser.rs]: align - typo,
thinked it better always add comma beetwen cases.

**Comment by EPashkin** [comment]
> filter(!is_empty()) in main crates' build scripts is needed to avoid returning an empty cfg that causes an ICE in rustc.

Thinked that empty part of `split` result can be only from empty string which better check directly. Now see other option ("3.0.").

> Regarding the version numbers complexity. Do you think there's any harm in using a tuple and defaulting failed elements to 0?

I plan replace empty parts with 0. `"3.1." => Version([3,1,0])` `"3..2" => Version([3,0,2])`.
Don't thinked about tupples.
Possible problems:
- comparing different length (2,3) and (1,4,2),
- iterating (mostly on comparing)

**Comment by gkoz** [comment]
If we decide to care about the patch version part, iterating blindly doesn't seem possible, we'll have to use the actual set of versions encountered when parsing the library (have `Library` maintain that set).

A fixed length tuple that defaults to zero seems an easy compromise. To rephrase the linked example,
```rust
    let mut parts = lib.version.splitn(4, '.')
        .map(|s| s.parse())
        .take_while(Result::is_ok)
        .map(Result::unwrap);
    let version: (u16, u16, u16) = (parts.next().unwrap_or(0), parts.next().unwrap_or(0), parts.next().unwrap_or(0));
```
If some of the parts fails to parse the tuple is filled with zeroes (`"3.1."` => `(3, 1, 0)`, `"2.a.1"` => `(2, 0, 0)`). Implementing `Ord` is straightforward, formatting `cfg` directives is manageable.

**Comment by EPashkin** [comment]
Queston, we need about add patch version in cfgs or not?

**Comment by gkoz** [comment]
Let's only add them if they're non-zero: `pango_1_32_4`, `pango_1_34`.

**Comment by EPashkin** [comment]
Ok.

**Comment by EPashkin** [comment]
And add warning to log this.

**Comment by gkoz** [comment]
The more specific versions don't seem to have any impact that would justify a warning.

**Comment by EPashkin** [comment]
But we know, that in library some function has really 3digits version, and adjust decision about iterating in build.rs

**Comment by EPashkin** [comment]
Remake Version to tuple

File: src/versions.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+use std::str::FromStr;
+
+#[derive(Debug, Clone, PartialEq)]
+// major, minor, patch
+pub struct Version(pub (u16,u16,u16));
```
Comment (gkoz) [inline src/versions.rs]: Could this be `Version(pub u16, pub u16, pub u16)`?

**Comment by EPashkin** [comment]
We have 1.32.4 in pango and some odd minored in others. Seems we really need store version in Namespace, and use it when generating build.rs or include to it  module with versions.

File: src/versions.rs (hunk 2 above)
Comment (EPashkin) [inline src/versions.rs]: Yes :(

File: src/main.rs [hunk 3]
```diff
@@ -24,6 +24,7 @@ mod nameutil;
 mod parser;
 mod traits;
 mod version;
+mod versions;
```
Comment (gkoz) [inline src/main.rs]: I'd prefer a useful module use the name `version` and the former `version` renamed to something else.

**Comment by EPashkin** [comment]
Why u16, not u32?

**Comment by EPashkin** [comment]
like `gir_version` ?

**Comment by gkoz** [comment]
> Why u16, not u32?

`u16` seemed exceedingly big for any sane version numbers.

> like `gir_version` ?

Sounds good.

File: src/versions.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+impl FromStr for Version {
+    type Err = String;
+
+    /// Parse a `Version` from a string.
+    fn from_str(s: &str) -> Result<Version, String> {
+        let mut parts = s.splitn(4, '.')
+            .map(|s| s.parse())
+            .take_while(Result::is_ok)
+            .map(Result::unwrap);
+        let version: (u16, u16, u16) = (parts.next().unwrap_or(0),
```
Comment (gkoz) [inline src/versions.rs]: Now you can stick the name `Version` in here and avoid enumerating the fields below.

**Comment by EPashkin** [comment]
Renamed files

File: src/version.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+use std::str::FromStr;
+
+#[derive(Debug, Clone, PartialEq)]
```
Comment (gkoz) [inline src/version.rs]: The `check_version` signature got me thinking: `Copy` should go here too.

**Comment by gkoz** [comment]
Looks good. Do you want to merge or build on this more?

**Comment by EPashkin** [comment]
As classes and interfaces seems not versioned, I only add deprecated-version for functions

**Comment by gkoz** [comment]
We might also be interested in enumeration and bitfield members' versions.

**Comment by EPashkin** [comment]
And constants.
IMHO better add config check to function before this, so ready for merge.
Warnings about versions I have not remove, `check_version` will changed to store versions to namespace. But it can be done only after generating main code to prevent desynchronization.
Or you prefer remove `check_version` now?

**Comment by EPashkin** [comment]
With version autodetection we have problems if generate config check on sys before do it on full generated main :cry:

**Comment by gkoz** [comment]
> Or you prefer remove check_version now?

It's fine for now.

Re. desynchronization, do I understand it correctly that your concern is about compatibility between the 'sys' and 'normal' modes because the 'normal' mode doesn't know about cfg versions? The objective of replacing manual sys crates needs the existence of cfg versions but doesn't require any conditional compilation in sys crates.

**Comment by gkoz** [comment]
@homu r+

**Comment by EPashkin** [comment]
Seems I don't understand something.
We have sys with autodetected `cfg_xxx` and function
```
#[cfg(gtk_3_12)]
 pub fn GTK_POPOVER(widget: *mut ffi::GtkWidget) -> *mut ffi::GtkPopover {
     unsafe { ffi::cast_GtkPopover(widget) }
 }
```
its current analog `gtk_popover_get_type` with same cfg. Or `gtk_popover_new`.
normal gtk get cfg value through DEP_GTK_CFG.
My concern is that:
if `pub use self::popover::Popover;` in gtk main.rs don't have same `#[cfg(gtk_3_12)]` for some reason,
then it not detected on CI as it use latest version.

**Comment by gkoz** [comment]
`main.rs` suggests an application? Testing applications for compatibility with older libraries could be difficult.
Our CI currently builds with both GTK 3.6 (travis and appveyor) and 3.14 (appveyor only), so missing a `cfg` should get caught.

**Comment by EPashkin** [comment]
Oh, forgot that gtk default is 3.6 - almost oldest supported version.

**Comment by EPashkin** [comment]
Then no obvious problems, but still don't like that condition on sys may added before same condition on normal.

**Pull request merged**